### ISSUE TITLE
Percolator: Could not convert string '8Plus'

### ISSUE THREAD

**Issue by wsnoble** [crux-toolkit/crux-toolkit#582]
I am trying to run crux percolator on a file generated by Kojak.  This file runs correctly with the stand-alone version of Percolator. But in Crux, it runs *nearly* to completion and then dies with the error `Could not convert string '8Plus'`.  This is the name of the last feature in the input file, which makes me think Crux is trying to parse the Percolator outputs.

I am attaching the pin file (which I have to name .txt to in order to upload) and the log file.

[XLpeplib_Beveridge_QEx-HFX_DSS_R1.perc.intra.txt](https://github.com/crux-toolkit/crux-toolkit/files/9172090/XLpeplib_Beveridge_QEx-HFX_DSS_R1.perc.intra.txt)
[percolator.log.txt](https://github.com/crux-toolkit/crux-toolkit/files/9172091/percolator.log.txt)

**Comment by wsnoble** [comment]
Note that this happens at the very end, after Percolator thinks it's finished.  I don't know why Crux should be trying to parse the PSM file at this point.

```
INFO: Found 989 test set PSMs with q<0.01.
INFO: Selected best-scoring PSM per scan+expMass (target-decoy competition): 13\
69 target PSMs and 322 decoy PSMs.
INFO: Calculating q values.
INFO: Final list yields 965 target PSMs with q<0.01.
INFO: Calculating posterior error probabilities (PEPs).
INFO: Processing took 4.0400 cpu seconds or 2 seconds wall clock time.
FATAL: An exception occurred: Could not convert string '8Plus'
```

**Comment by wsnoble** [comment]
The answer is that Crux is trying to produce pepxml or mzidentml files.  This will clearly fail for a non-standard file like this.  I will try to make Crux more robust to this case.

**Comment by wsnoble** [comment]
OK, I need help.  In the patch below, I tried to refactor the end of PercolatorApplication.cpp so that it would only attempt to parse the Percolator output files if the user requested mzid or pepxml files. This gets rid of the '8plus' error message mentioned in the title of this issue.  However, it still yields an error, and I can't seem to figure out where it's coming from. 

Here are the final few lines of the output from Percolator:
```
INFO: Calculating q values.
INFO: Final list yields 965 target PSMs with q<0.01.
INFO: Calculating posterior error probabilities (PEPs).
INFO: Processing took 3.9600 cpu seconds or 2 seconds wall clock time.
FATAL: An exception occurred: Invalid character '(' in sequence '-.MDGTEELLVKLNR(10)--MDGTEELLVKLNR(10).-'
```

Note that the "Processing took" line indicates that Percolator is completely finished.  In the patch, you'll see that I've added a CARP statement immediately thereafter to say 

```
carp(CARP_INFO, "Finished Percolating!");
```

The program never gets there.  However, somehow the message above gets generated.  It's coming from Modification.cpp at line 625. I can't for the life of me figure out how it's happening, partly because the whole "Caller.h" thing for PercolatorApplication.cpp is mysterious.  Line 8 of that file does this:

```
#include "Caller.h"
```

but I can find no evidence of any such file anywhere.  I simply want Crux to call Percolator on this file and then be done, and it insists on trying to read the file. Rgh!

Here is the full command line I am using:

```
crux percolator \
    --only-psms T \
    --tdc T \
    --decoy-prefix DECOY_ \
    --overwrite T \
    XLpeplib_Beveridge_QEx-HFX_DSS_R1.perc.intra.pin
```
[percfix.txt](https://github.com/crux-toolkit/crux-toolkit/files/9172570/percfix.txt)

**Comment by CharlesEGrant** [comment]
caller.h is part of the percolator source. It should be found in ext/build/src/percolator/src/Caller.h in the build directory.

**Comment by wsnoble** [comment]
OK, I think I came up with a solution here.  The basic problem is that Crux's PSM parsing code can't parse PSMs from Kojak, since they contain cross-linked peptides, which Crux knows nothing about.  However, the key insight is that if the user doesn't ask for pepxml or mzid output, then there is no reason for Crux to parse those PSMs in the first place.  I therefore refactored the code a bit so that it would check whether either of those two output formats was requested, by inserting a line like this in various places:

```
 if ( Params::GetBool("mzid-output") || Params::GetBool("pepxml-output") ) {
```

The upshot is that now Percolator will run to completion unless the user asks for one of those two output formats.  It still dies ungracefully, but I think that's OK for now. I anticipate that we may simply remove these output formats from Percolator in a future version.

Note that I also fixed a small bug in which the peptide-level output files were being requested even if the user specified the --only-psms option.
